File: 1717.cpp
#include<vector>
#include<stdio.h>

using namespace std;

int numOfVertices, numOfInputs;
vector<int> group;

void input();
void solve();
int find(int a);
void merge(int a, int b);
void test(int a, int b);

int main() {
    input();
    solve();
    return 0;
}

void input() {
    scanf("%d %d", &numOfVertices, &numOfInputs);
    for(int i = 0; i <= numOfVertices; i++) {
        group.push_back(i);
    }
}

void solve() {
    for(int i = 0; i < numOfInputs; i++) {
        int command, from, to;
        scanf("%d %d %d", &command,  &from, &to);
        
        if(command == 0) {
            merge(from, to);
        } else if(command == 1) {
            test(from, to);
        }
    }
}

int find(int a) {
    if(group[a] == a) {
        return a;
    } else {
        return group[a] = find(group[a]);
    }
}

void merge(int a, int b) {
    int A = find(a);
    int B = find(b);
    
    if(A != B) {
        if(B < A)
            group[A] = B;
        else
            group[B] = A;
    }
}

void test(int a, int b) {
    int A = find(a);
    int B = find(b);
    
    if(A == B)
        printf("YES\n");
    else
        printf("NO\n");
}

File: 1717.swift
import Foundation

let input = readLine()!.split(separator: " ").map({Int(String($0))!})
let numOfVertices = input[0]
let numOfInputs = input[1]

var actualGroup = Array(repeating: -1, count: numOfVertices + 1)

for i in 0...numOfVertices {
    actualGroup[i] = i
}

for _ in 0..<numOfInputs {
    let eachCommand = readLine()!.split(separator: " ").map({Int(String($0))!})
    let command = eachCommand[0]
    let from = eachCommand[1]
    let to = eachCommand[2]
    
    if command == 0 {
        union(from, to)
    } else if command == 1 {
        isInSameGroup(from, to)
    }
}

func find(_ vertex: Int) -> Int {
    if actualGroup[vertex] == vertex {
        return vertex
    } else {
        actualGroup[vertex] = find(actualGroup[vertex])
        return actualGroup[vertex]
    }
}

func union(_ a: Int, _ b: Int) {
    let actualA = find(a)
    let actualB = find(b)
    
    if actualA != actualB {
        actualGroup[actualB] = actualA
    }
}

func isInSameGroup(_ a: Int, _ b: Int) {
    let actualA = find(a)
    let actualB = find(b)
    
    if actualA == actualB {
        print("YES")
        return
    }
    print("NO")
}

